src/storage/buffer_pool_manager/buffer_pool.rs

```diff
@@ -0,0 +1,432 @@
+use dashmap::DashMap;
+use parking_lot::{lock_api::RawRwLock, Mutex, RawRwLock as PLRawRwLock, RwLock};
+use std::{cell::UnsafeCell, ops::Range};
+
+pub mod contracts {
+    use super::*;
+
+    pub struct PageReadLock<'a> {
+        page_table_item_read_lock_raw: &'a PLRawRwLock,
+        frame_read_lock_raw: &'a PLRawRwLock,
+        pub data: &'a [u8],
+    }
+
+    impl<'a> PageReadLock<'a> {
+        pub(super) fn new(
+            page_table_item_read_lock_raw: &'a PLRawRwLock,
+            frame_read_lock_raw: &'a PLRawRwLock,
+            data: &'a [u8],
+        ) -> Self {
+            Self {
+                page_table_item_read_lock_raw,
+                frame_read_lock_raw,
+                data,
+            }
+        }
+    }
+
+    impl<'a> std::fmt::Debug for PageReadLock<'a> {
+        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            f.debug_struct("PageReadLock")
+                .field("data", &self.data)
+                .finish()
+        }
+    }
+
+    impl<'a> Drop for PageReadLock<'a> {
+        fn drop(&mut self) {
+            // Safety: We assume that the locks were created properly.
+            unsafe {
+                // Safety: We unlock the inner lock first.
+                self.frame_read_lock_raw.unlock_shared();
+                self.page_table_item_read_lock_raw.unlock_shared();
+            }
+        }
+    }
+
+    pub type BufferPoolKey = (TableId, PageId);
+
+    pub trait BufferPool: Send + Sync + 'static {
+        fn get_page(&self, id: BufferPoolKey) -> PageReadLock;
+        /// # Safety
+        /// The buffer pool must contain `id`.
+        /// `offset % page_size == 0`.
+        /// `offset + data.len() <= page_size`.
+        fn mutate_page(&self, id: BufferPoolKey, offset: usize, data: Vec<u8>);
+    }
+}
+
+#[derive(Debug)]
+#[repr(u8)]
+/* private */
+enum PageTableItemMetadataFlagsMasks {
+    IsDirty = 0b00000001,
+}
+
+#[derive(Default, Debug)]
+#[repr(transparent)]
+/* private */
+struct PageTableItemMetadataFlags(u8);
+
+macro_rules! create_page_table_item_metadata_flags_get_set {
+    ($mask: expr, $get_name: ident, $set_name: ident) => {
+        pub const fn $get_name(&self) -> bool {
+            (self.0 & ($mask as u8)) != 0
+        }
+
+        pub fn $set_name(&mut self, value: bool) {
+            self.0 = if value {
+                self.0 | ($mask as u8)
+            } else {
+                self.0 | (!($mask as u8))
+            };
+        }
+    };
+}
+
+impl PageTableItemMetadataFlags {
+    create_page_table_item_metadata_flags_get_set!(
+        PageTableItemMetadataFlagsMasks::IsDirty,
+        get_is_dirty,
+        set_is_dirty
+    );
+}
+
+#[derive(Default, Debug)]
+/* private */
+struct PageTableItemMetadata {
+    access_count: usize,
+    pin_count: usize,
+    flags: PageTableItemMetadataFlags,
+}
+
+#[derive(Debug)]
+/* private */
+enum PageTableItem {
+    Free {
+        next_free_index: Mutex<Option<usize>>,
```

Comment:
Same here - I believe this can also just be atomic u32 (I don't think we should be using usize so extensively since the # of bytes is platform dependent)